### DISCUSSION TITLE
SIDE_MODULE and USE_PTHREADS provoke memory growth

### DISCUSSION THREAD

**Discussion by alexander-veselov** [emscripten-core/emscripten#17034] [Q&A] [answered]
I understand that "SIDE_MODULE + USE_PTHREADS" is experimental. But I would still like to know if this is a known issue and if it can be fixed somehow.

We have a lot of side modules (about 350MB). USE_PTHREADS=1 and PTHREAD_POOL_SIZE=20. With this configuration, an "out of memory" error occurs. It seems that all side modules are loaded several times for each thread.

For example, if you set PTHREAD_POOL_SIZE=1, then the problem will disappear.
I know that when building without side modules, the problem will also disappear, but we need side modules.

Is it possible to make threads use shared memory and not load copies of dynamic libs several times?

**Comment by kripken** [comment, marked as answer]
@sbc100 Is it the case that loading a dynamic library compiles on that pthread, so each pthread ends up recompiling the same code? (Unlike with the main module, which we compile once and just send that to the workers.)

I'm not sure how we could improve that. To send the compiled dynamic library we'd need to do an async `postMessage`. Perhaps with Asyncify or stack switching we could pause and wait for that.

**Reply by alexander-veselov** [reply]
We decided not to use Asyncify because of runtime performance hit, so we would need some solution that doesn't involve Asyncify.

**Reply by sbc100** [reply]
Another solution would be to move as much of your work as possible onto a background worker, where blocking is allowed (or at least more tolerated).   For example, `PROXY_TO_PTHREAD` buts your `main()` function onto a thread.

**Reply by alexander-veselov** [reply]
Blocking is not a problem for us. The main problem is memory consumption. With an increase in the number of threads, the memory size increases in accordance with the size of the side modules. So what we want to know is, is it possible to share a module across multiple threads?

**Reply by sbc100** [reply]
It could be possible yes.  With the main module we use post message to send the module between workers but we don't currently do that for side modules.

Can you open a bug/feature request to investigate?  (Or we can convert this discussion into a bug).

Can I ask, are you module loaded at startup or at runtime using dlopen?

**Reply by alexander-veselov** [reply]
Currently we load modules at startup. But in general we plan to use both approaches.

**Reply by sbc100** [reply]
Thats good.  The startup use-case should easier to handle. since in the cast the main thread can take care of postMessaging all the modules over to newly started threads because they actually start running. 

For the dlopen case its tricky because the library should be loaded by any thread.. and dlopen is expected to by synchronous so we can't always return the even loop to process messages in the middle of it.

**Reply by alexander-veselov** [reply]
It would acceptable for us even if only main thread can use dlopen.

**Reply by sbc100** [reply]
The problem is that once the main thread calls dlopen there is a general expectation that the code can run on all threads.  We have tried to our best to keep all threads in sync.. but its tricky.   See https://github.com/emscripten-core/emscripten/blob/b2dc6cfad086059bec9b713a8bd0ac313f559fd6/system/lib/libc/dynlink.c#L221-L224

Would you would like to turn this discussion into a feature request, or open an new issue "Use postMessage to share side modules"?   I'm not sure when I will get a chance to work on it but it might be something someone else want to a take a stab at.

**Reply by alexander-veselov** [reply]
I'm not sure. What's the difference? Well, let it be issue

**Reply by sbc100** [reply]
New issue: https://github.com/emscripten-core/emscripten/issues/17078
